File: jelly-vs-tower/python/src/jelly_vs_tower/arena.py
from __future__ import annotations

from .combat_log import CombatLog
from .damage_table import calculate_damage
from .jelly import Jelly
from .random_source import RandomSource
from .tower import Tower


class Arena:
    def __init__(
        self,
        towers: list[Tower],
        jellies: list[Jelly],
        random: RandomSource,
    ) -> None:
        self._towers = list(towers)
        self._jellies = list(jellies)
        self._random = random

    @property
    def alive_jellies(self) -> list[Jelly]:
        return [j for j in self._jellies if j.is_alive]

    def execute_round(self) -> list[CombatLog]:
        logs: list[CombatLog] = []

        for tower in self._towers:
            target = next((j for j in self._jellies if j.is_alive), None)
            if target is None:
                break

            damage = calculate_damage(tower, target, self._random)
            target.take_damage(damage)
            logs.append(CombatLog(tower.id, target.id, damage))

        return logs

File: jelly-vs-tower/typescript/src/Arena.ts
import { CombatLog } from './CombatLog.js';
import { calculateDamage } from './DamageTable.js';
import { Jelly } from './Jelly.js';
import { RandomSource } from './RandomSource.js';
import { Tower } from './Tower.js';

export class Arena {
  private readonly towers: readonly Tower[];
  private readonly jellies: readonly Jelly[];
  private readonly random: RandomSource;

  constructor(towers: readonly Tower[], jellies: readonly Jelly[], random: RandomSource) {
    this.towers = towers;
    this.jellies = jellies;
    this.random = random;
  }

  get aliveJellies(): Jelly[] {
    return this.jellies.filter(j => j.isAlive);
  }

  executeRound(): CombatLog[] {
    const logs: CombatLog[] = [];

    for (const tower of this.towers) {
      const target = this.jellies.find(j => j.isAlive);
      if (!target) break;

      const damage = calculateDamage(tower, target, this.random);
      target.takeDamage(damage);
      logs.push({ towerId: tower.id, jellyId: target.id, damage });
    }

    return logs;
  }
}
